vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -0,0 +1,476 @@
+                time_stamp = entry.time_stamp;
+                phase = entry.phase;
+                accumulated_message.clear();
+                is_accumulating = true;
+            }
+            accumulated_message.push_str(message);
+
+            // Validate that the accumulated message is not too long
+            if accumulated_message.len() > MAX_MESSAGE_LENGTH as usize {
+                return Err(DiagnosticsError::GeneralError(format!(
+                    "Accumulated message length ({}) exceeds max length ({})",
+                    accumulated_message.len(),
+                    MAX_MESSAGE_LENGTH
+                )));
+            }
+
+            // Completed messages (ending with '\n') become log entries
+            if !message.is_empty() && message.ends_with('\n') {
+                logs.push(EfiDiagnosticsLog {
+                    debug_level,
+                    time_stamp,
+                    phase,
+                    message: std::mem::take(&mut accumulated_message)
+                        .trim_end_matches(&['\r', '\n'][..])
+                        .to_string(),
+                });
+                entries_processed += 1;
+                is_accumulating = false;
+            }
+
+            //
+            // Step 5c: Move to the next entry
+            //
+
+            // Calculate base offset (entry header size + message length)
+            let base_offset = size_of::<AdvancedLoggerMessageEntryV2>()
+                .checked_add(message_len as usize)
+                .ok_or_else(|| {
+                    DiagnosticsError::Overflow(format!(
+                        "size_of::<AdvancedLoggerMessageEntryV2> ({}) + message_len ({})",
+                        size_of::<AdvancedLoggerMessageEntryV2>(),
+                        message_len
+                    ))
+                })?;
+
+            // Add padding for 8-byte alignment
+            let aligned_offset = base_offset.checked_add(ALIGNMENT_MASK).ok_or_else(|| {
+                DiagnosticsError::Overflow(format!(
+                    "base_offset ({}) + ALIGNMENT_MASK ({})",
+                    base_offset, ALIGNMENT_MASK
+                ))
+            })?;
+            let next_offset = aligned_offset & !ALIGNMENT_MASK;
+
+            // Update overall bytes read counter
+            bytes_read = bytes_read.checked_add(next_offset).ok_or_else(|| {
+                DiagnosticsError::Overflow(format!(
+                    "bytes_read ({}) + next_offset ({})",
+                    bytes_read, next_offset
+                ))
+            })?;
+
+            // Advanced to the next entry with boundary checks
+            if next_offset >= buffer_slice.len() {
+                // We have reached the end of the buffer
+                break;
+            }
+            buffer_slice = &buffer_slice[next_offset..];
+        }
+
+        // Process remaining messages
+        if is_accumulating {
+            logs.push(EfiDiagnosticsLog {
```

Comment:
If we're just logging these, why do we need to put them into a Vec first?

Can we instead have the caller pass an `FnMut(EfiDiagnosticsLog)`, to avoid this?